# Like ESLint core yoda rules but with more options (yoda-ex)

Please describe the origin of the rule here.


## Rule Details

This rule aims to...

Examples of **incorrect** code for this rule:

```js

// fill me in

```

Examples of **correct** code for this rule:

```js

// fill me in

```

### Options

If there are any options, describe them here. Otherwise, delete this section.

## When Not To Use It

Give a short description of when it would be appropriate to turn off this rule.

## Further Reading

 - [ESLint core yoda rule](https://eslint.org/docs/rules/yoda)
 - [Yoda Conditions](https://en.wikipedia.org/wiki/Yoda_conditions)
 - [Yoda Notation and Safe Switching](http://thomas.tuerke.net/on/design/?with=1249091668#msg1146181680)
